libp2p/NodeTable.cpp

```diff
@@ -88,62 +86,83 @@ void NodeTable::processEvents()
         m_nodeEventHandler->processEvents();
 }
 
-bool NodeTable::addNode(Node const& _node, NodeRelation _relation)
+bool NodeTable::addNode(Node const& _node)
+{
+    shared_ptr<NodeEntry> entry;
+    DEV_GUARDED(x_nodes)
+    {
+        auto const it = m_allNodes.find(_node.id);
+        if (it == m_allNodes.end())
+            entry = createNodeEntry(_node, 0, 0);
+        else
+            entry = it->second;
+    }
+    if (!entry)
+        return false;
+
+    if (!entry->hasValidEndpointProof())
+        ping(*entry);
+
+    return true;
+}
+
+bool NodeTable::addKnownNode(
+    Node const& _node, uint32_t _lastPongReceivedTime, uint32_t _lastPongSentTime)
+{
+    shared_ptr<NodeEntry> entry;
+    DEV_GUARDED(x_nodes)
+    {
+        entry = createNodeEntry(_node, _lastPongReceivedTime, _lastPongSentTime);
+    }
+    if (!entry)
+        return false;
+
+    if (entry->hasValidEndpointProof())
+        noteActiveNode(entry->id, entry->endpoint);
+    else
+        ping(*entry);
+
+    return true;
+}
+
+std::shared_ptr<NodeEntry> NodeTable::createNodeEntry(
+    Node const& _node, uint32_t _lastPongReceivedTime, uint32_t _lastPongSentTime)
 {
     if (!_node.endpoint || !_node.id)
     {
         LOG(m_logger) << "Supplied node has an invalid endpoint (" << _node.endpoint << ") or id ("
```

Comment:
Logs in this function are now executed under `x_nodes` mutex, but I don't think it's a big problem, as logging is now asynchronous (in Release build), and also the part to the right of `LOG(...) << ` is executed only when not disabled by log filters.